IPython/html/static/notebook/js/kernelselector.js

```diff
@@ -154,28 +160,116 @@ define([
+            };
+        }
+        if (this._loaded) {
+            this._set_kernel(selected);
+        } else {
+            return this.loaded.then(function () {
+                that._set_kernel(selected);
+            });
+        }
     };
 
-    KernelSelector.prototype._set_kernel = function (kernel_name) {
+    KernelSelector.prototype._set_kernel = function (selected) {
         /** Actually set the kernel (kernelspecs have been loaded) */
-        if (kernel_name === this.current_selection) {
+        if (selected.name === this.current_selection) {
             // only trigger event if value changed
             return;
         }
-        var ks = this.kernelspecs[kernel_name];
+        var kernelspecs = this.kernelspecs;
+        var ks = kernelspecs[selected.name];
+        if (ks === undefined) {
+            var available = _sorted_names(kernelspecs);
+            var matches = [];
+            if (selected.language && selected.language.length > 0) {
+                available.map(function (name) {
+                    if (kernelspecs[name].spec.language.toLowerCase() === selected.language.toLowerCase()) {
+                        matches.push(name);
+                    }
+                });
+            }
+            if (matches.length === 1) {
+                ks = kernelspecs[matches[0]];
+                console.log("No exact match found for " + selected.name +
+                    ", using only kernel that matches language=" + selected.language, ks);
+                this.events.trigger("spec_match_found.Kernel", {
+                    selected: selected,
+                    found: ks,
+                });
+            }
+            // if still undefined, trigger failure event
+            if (ks === undefined) {
+                if (matches) {}
+                this.events.trigger("spec_not_found.Kernel", {
+                    selected: selected,
+                    matches: matches,
+                    available: available,
+                });
+                return;
+            }
+        }
         if (this.notebook._session_starting) {
             console.error("Cannot change kernel while waiting for pending session start.");
             return;
         }
-        this.current_selection = kernel_name;
+        this.current_selection = ks.name;
         this.events.trigger('spec_changed.Kernel', ks);
     };
+    
+    KernelSelector.prototype._spec_not_found = function (event, data) {
+        var that = this;
+        var select = $("<select>").addClass('form-control');
+        console.warn("Kernelspec not found:", data);
+        var names;
+        if (data.matches.length > 1) {
+            names = data.matches;
+        } else {
+            names = data.available;
+        }
+        names.map(function (name) {
+            var ks = that.kernelspecs[name];
+            select.append(
+                $('<option/>').attr('value', ks.name).text(ks.spec.display_name || ks.name)
```

Comment:
Because display_name is part of a kernelspec, the spec's name (the folder in which it's contained) is not part of the spec itself.